packages/switch/index.js

```diff
@@ -0,0 +1,91 @@
+import React, {Component} from 'react';
+import classnames from 'classnames';
+import PropTypes from 'prop-types';
+import {MDCSwitchFoundation} from '@material/switch/dist/mdc.switch';
+
+import ThumbUnderlay from './ThumbUnderlay';
+
+export default class Switch extends Component {
+  foundation_ = null;
+  state = {
+    checked: this.props.checked,
+    classList: new Set(),
+    disabled: this.props.disabled,
+  };
+
+  componentDidMount() {
+    this.foundation_ = new MDCSwitchFoundation(this.adapter);
+    this.foundation_.init();
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    if (this.state.checked !== prevState.checked) {
+      this.foundation_.handleChange();
+    }
+  }
+
+  componentWillUnmount() {
+    this.foundation_.destroy();
+  }
+
+  get classes() {
+    const {classList} = this.state;
+    const {className, disabled} = this.props;
+    return classnames('mdc-switch', Array.from(classList), className, {
+      'mdc-switch--disabled': disabled,
+    });
+  }
+
+  get adapter() {
+    return {
+      addClass: (className) => {
+        const {classList} = this.state;
+        classList.add(className);
+        this.setState({classList});
+      },
+      removeClass: (className) => {
+        const {classList} = this.state;
+        classList.delete(className);
+        this.setState({classList});
+      },
+      isNativeControlChecked: () => this.state.checked,
```

Comment:
is it weird that we implement `isNativeControlChecked` by look at the root component's state? I guess we are basically guaranteeing that if the native control's state changes, then ThumbUnderlay will respond in an `onChange` callback. And then if ThumbUnderlay's state changes we guarantee that Switch's state will change also with and `handleChange` callback. 

But for some reason that feels a bit fragile. Did we do the same design for the text field?

Reply:
Yep, but text field only has one layer of nesting (is nesting the right word?) `TextField` -> `Input` which makes it seem less fragile. And similarly `Select`->`NativeControl`. Switch is the only component for now with 3 layers because it needs one for the native input, one for the ripple surface, and one for the top-level component.